hs/src/Reach/Eval.hs

```diff
@@ -1401,6 +1402,106 @@ evalPrimOp ctxt at _sco st p sargs =
     BXOR -> nn2n (xor)
     SELF_ADDRESS -> impossible "self address"
   where
+    polyEq args' =
+      case args' of
+        -- Both args static
+        [SLV_Int _ l, SLV_Int _ r]      -> retBool $ l == r
+        [SLV_Bool _ l, SLV_Bool _ r]    -> retBool $ l == r
+        [SLV_Bytes _ l, SLV_Bytes _ r]  -> retBool $ l == r
+        [SLV_Type l, SLV_Type r]        -> retBool $ l == r
+        [SLV_Null {}, SLV_Null {}]      -> retBool True
+        [SLV_Array lAt lTy ls, SLV_Array rAt rTy rs]
+          | lTy /= rTy || lLen /= rLen ->
+            expect_throw_ctx ctxt at $ Err_TypeMeets_Mismatch at
+              (lAt, T_Array lTy $ toInteger lLen) (rAt, T_Array rTy $ toInteger rLen)
+          | otherwise -> staticCmpElems ls rs
+          where
+            lLen = length ls
+            rLen = length rs
+        [SLV_Tuple lAt ls, SLV_Tuple rAt rs]
+          | lTy /= rTy ->
+            expect_throw_ctx ctxt at $
+              Err_TypeMeets_Mismatch at (lAt, lTy) (rAt, rTy)
+          | otherwise -> staticCmpElems ls rs
+          where
+            lTy = T_Tuple $ map getType ls
+            rTy = T_Tuple $ map getType rs
+        [l@(SLV_Object lAt _ lEnv), r@(SLV_Object rAt _ rEnv)]
+          | S.difference (getObjFieldTypes lEnv) (getObjFieldTypes rEnv) /= S.empty ->
+            expect_throw_ctx ctxt at $
+              Err_TypeMeets_Mismatch at (lAt, getType l) (rAt, getType r)
+          | otherwise -> do
+            let elems = map sss_val . M.elems
+            staticCmpElems (elems lEnv) $ elems rEnv
+          where
+            getObjFieldTypes xs = S.fromAscList $ map getFieldType $ M.toAscList xs
+            getFieldType (k, v) = (k, getType $ sss_val v)
+        [SLV_Data lAt lCons lCon lVal, SLV_Data rAt rCons rCon rVal]
+          | S.difference (S.fromAscList $ M.toAscList lCons)
+            (S.fromAscList $ M.toAscList rCons) /= S.empty ->
+              expect_throw_ctx ctxt at $
+                Err_TypeMeets_Mismatch at (lAt, T_Data lCons) (rAt, T_Data rCons)
+          | lCon /= rCon -> retBool False
+          | otherwise    -> polyEq [lVal, rVal]
+        -- If atleast one arg is dynamic
+        [l, r] ->
+          case (getType l, getType r) of
+            (T_UInt, T_UInt)        -> make_var
+            (T_Null, T_Null)        -> retBool True
+            (T_Digest, T_Digest)    -> make_var
+            (T_Address, T_Address)  -> make_var
+            (T_Bool, T_Bool) -> do
+              let fn = sss_val $ env_lookup (Just ctxt) at (LC_RefFrom "polyEq") "boolEq" (sco_env _sco)
+              SLRes cmp_lifts _ (SLAppRes _ val) <- evalApplyVals ctxt at _sco st fn $ map (\ a -> (lvl, a)) args
+              return $ SLRes cmp_lifts st val
+            (T_Bytes l_len, T_Bytes r_len)
```

Comment:
I think this case, and all the ones after it, should be same code path and should use `typeMeet` or `typeEqual` something to ensure that the types are the same using the code in Type.hs and then just do the hash that you have